hpy/debug/src/_debugmod.c

```diff
@@ -0,0 +1,212 @@
+
+// NOTE: hpy.debug._debug is loaded using the UNIVERSAL ctx. To make it
+// clearer, we will use "uctx" and "dctx" to distinguish them.
+
+#include "hpy.h"
+#include "debug_internal.h"
+
+static UHPy new_DebugHandleObj(HPyContext uctx, UHPy u_DebugHandleType,
+                               DebugHandle *handle);
+
+
+HPyDef_METH(new_generation, "new_generation", new_generation_impl, HPyFunc_NOARGS)
+static UHPy new_generation_impl(HPyContext uctx, UHPy self)
+{
+    HPyContext dctx = hpy_debug_get_ctx(uctx);
+    HPyDebugInfo *info = get_info(dctx);
+    info->current_generation++;
+    return HPyLong_FromLong(uctx, info->current_generation);
+}
+
+
+HPyDef_METH(get_open_handles, "get_open_handles", get_open_handles_impl, HPyFunc_O, .doc=
+            "Return a list containing all the open handles whose generation is >= "
+            "of the given arg")
+static UHPy get_open_handles_impl(HPyContext uctx, UHPy u_self, UHPy u_gen)
+{
+    HPyContext dctx = hpy_debug_get_ctx(uctx);
+    HPyDebugInfo *info = get_info(dctx);
+
+    UHPy u_DebugHandleType = HPy_GetAttr_s(uctx, u_self, "DebugHandle");
+    if (HPy_IsNull(u_DebugHandleType))
+        return HPy_NULL;
+
+    long gen = HPyLong_AsLong(uctx, u_gen);
+    if (HPyErr_Occurred(uctx))
+        return HPy_NULL;
+
+    UHPy u_result = HPyList_New(uctx, 0);
+    if (HPy_IsNull(u_result))
+        return HPy_NULL;
+
+    DebugHandle *dh = info->open_handles;
+    while(dh != NULL) {
+        if (dh->generation >= gen) {
+            UHPy u_item = new_DebugHandleObj(uctx, u_DebugHandleType, dh);
+            if (HPy_IsNull(u_item))
+                return HPy_NULL;
+            if (HPyList_Append(uctx, u_result, u_item) == -1) {
+                HPy_Close(uctx, u_item);
+                HPy_Close(uctx, u_result);
+                return HPy_NULL;
+            }
+            HPy_Close(uctx, u_item);
+        }
+        dh = dh->next;
+    }
+    return u_result;
+}
+
+/* ~~~~~~ DebugHandleType and DebugHandleObject ~~~~~~~~
+
+   This is the applevel view of a DebugHandle/DHPy.
+
+   Note that there are two different ways to expose DebugHandle to applevel:
+
+   1. make DebugHandle itself a Python object: this is simple but means that
+      you have to pay the PyObject_HEAD overhead (16 bytes) for all of them
+
+   2. make DebugHandle a plain C struct, and expose them through a
+      Python-level wrapper.
+
+   We choose to implement solution 2 because we expect to have many
+   DebugHandle around, but to expose only few of them to applevel, when you
+   call get_open_handles. This way, we save 16 bytes per DebugHandle.
+
+   This means that you can have different DebugHandleObjects wrapping the same
+   DebugHandle. To make it easier to compare them, they expose the .id
+   attribute, which is the address of the wrapped DebugHandle. Also,
+   DebugHandleObjects compare equal if their .id is equal.
+*/
+
+typedef struct {
+    HPyObject_HEAD
```

Comment:
Could we make this a pure type once the cast support branch lands?

Reply:
yes!

Reply:
Correct. HPyObject_HEAD is already gone in the cast support branch.